Implement peerCount JSON-RPC API

cburgdorf opened the pull request:
# THIS IS STILL WIP, NEEDS TESTS AND EATS KITTEN


### What was wrong?

The JSON-RPC API for `peerCount` wasn't implemented.

### How was it fixed?

Implemented it.

#### Cute Animal Picture

![put a cute animal picture link inside the parentheses]()

pipermerriam commented on trinity/rpc/modules/net.py:
```diff
@@ -10,3 +10,21 @@ def version(self):
+        Return the number of peers that are currently connected to the node
+        """
+        if self._p2p_server.peer_pool is None:
+            return '0x0'
```
I'm curious about this logic and why it's here.  Seems like the `net_` API should not be runnable if there is no peer pool.

cburgdorf replied:
Well, parts of the `net` module are independent from the peer pool (e.g. version) and by the time that the peer pool module is created, we don't have a peer pool yet. The peer pool is created asynchronously and it takes several seconds until it is set here: https://github.com/ethereum/py-evm/blob/f66d296074c0603eefa4e4c718421956d70d183e/p2p/server.py#L262

In other words, when you start trinity, the `net` module is already set up and answers to `net_peerCount` request and does therefore return `0x0` for when the `peer_pool` is still `None`. Which at least technically is correct :)

pipermerriam replied:
Hrm, can you open an issue to track this as I think that we should fix that.

pipermerriam replied:
Maybe if `peer_pool is None` should return a JSON-RPC error indicating that it's not ready?

pipermerriam replied:
> in fact our p2p subsystem is still bootstrapping and we have in fact 0 connected peers at that time.

👍

cburgdorf replied:
Wouldn't `hex(0)` cut it if you dislike the hardcoded string?

cburgdorf replied:
@pipermerriam I created #877 regarding the `PeerPool` issue.

pipermerriam commented on trinity/rpc/modules/net.py:
```diff
@@ -10,3 +10,21 @@ def version(self):
+        if self._p2p_server.peer_pool is None:
+            return False
+
+        return not self._p2p_server.peer_pool.is_full
```
Are we sure this is the right logic.  We are technically still listening (have an outbound socket open that accepts incoming connections) even when the peer pool is full.  I think the `discovery` service is where we'd need to look to determine if we're listening or not.

cburgdorf replied:
I'm gonna double check that. Here is what pyethapp does which isn't the same as what we do but not that far away either :sweat_smile: 

https://github.com/ethereum/pyethapp/blob/9af6e64e6140104ab13fdbc5a6b22a915c412adc/pyethapp/jsonrpc.py#L696

carver replied:
> `// (this may not be true now -- 26 september 2016)`

lol

cburgdorf commented on trinity/rpc/modules/main.py:
```diff
@@ -1,8 +1,9 @@
     _chain = None
 
-    def __init__(self, chain=None):
+    def __init__(self, chain=None, p2p_server=None):
```
[See my comment](https://github.com/ethereum/py-evm/pull/867#discussion_r193181748) here, by the time that the module is initialized the `peer_pool` isn't available yet

carver replied:
We might need it anyway, to add `eth_syncing`.

cburgdorf commented on p2p/peer.py:
```diff
@@ -582,7 +582,11 @@ def __init__(self,
+        return self.connected_peer_count >= self.max_peers
+
+    @property
+    def connected_peer_count(self) -> int:
```
Sounds pythonic to me :+1:

carver commented:
If you haven't yet, I would recommend checking if this works on the light client. I suspect there will be problems because it doesn't track the peer pool the same way.

cburgdorf commented:
@pipermerriam @carver 

Here's a summary of the state of this:

- I implemented `__len__` for `PeerPool` as suggested
- I refactored and added more tests. E.g. the tests now cover the case when the `PeerPool` is still `None` during bootstrapping etc.
- I tested the PR against `--light` and found the added APIs to be working fine

I'm still not sure what to do about the implementation logic for `net_listening`. Parity always returns `true`, `pyethapp` returns `true` if the number of peers is lower than the `min_peers`. I don't know what kind of implications our answer on this API endpoint has so for now it strikes me as balanced to say we are listening as long as we haven't reached the max cap of peers.

pipermerriam commented:
I think `net_listening` should be `true` when we are listening (even if the peer pool is full).

carver commented:
> I think net_listening should be true when we are listening (even if the peer pool is full).

I agree, [the spec says](https://github.com/ethereum/wiki/wiki/JSON-RPC#net_listening):
> `Returns true if client is actively listening for network connections.`

We are listening for connections even if full, just sending back a drop message.

Returning `False` before the p2p server has bound to the address/port makes sense, but I'm not sure how easy it is to get at that info, and it doesn't strike me as critical to implement right away.
